src/main/java/com/arpnetworking/metrics/mad/Aggregator.java

```diff
@@ -78,31 +80,29 @@ public synchronized void shutdown() {
                 .addData("aggregator", this)
                 .log();
 
-        final Semaphore barrier = new Semaphore(0);
-        int actorCount = 0;
+        final List<CompletableFuture<Object>> shutdownStages = new ArrayList<>();
         for (final List<ActorRef> actorRefList : _periodWorkerActors.values()) {
             for (final ActorRef actorRef : actorRefList) {
-                ++actorCount;
-                Patterns.ask(
-                        actorRef,
-                        PoisonPill.getInstance(),
-                        SHUTDOWN_TIMEOUT.toMillis())
-                        .andThen(
-                                new OnComplete<Object>() {
-                                    @Override
-                                    public void onComplete(final Throwable throwable, final Object result) {
-                                        barrier.release();
-                                    }
-                                },
-                        _actorSystem.getDispatcher());
+                shutdownStages.add(
+                        PatternsCS.ask(
+                                actorRef,
+                                PoisonPill.getInstance(),
+                                SHUTDOWN_TIMEOUT).toCompletableFuture());
+
+
             }
         }
         try {
-            barrier.acquire(actorCount);
+            CompletableFutures.allOf(shutdownStages).get();
```

Comment:
You might want an explicit timeout on this `get`.

I understand that the underlying futures all have their own timeouts, but my brain is already trained to ring alarm bells when I see a bare `get`.

Reply:
Sure. I can do that.